### PR TITLE
Refactor Location Permissions

### LINKED ISSUE #403: Refactor location permissions acording to android documentation
**Is your feature request related to a problem? Please describe.**
https://developer.android.com/training/permissions/requesting

### PR DESCRIPTION
**Pull request by maxammann**
fixes #403 

This PR changes how permissions for locations are handled. Previously we showed a Dialog which links into the such that the user can change the OS settings to reenable permissions. This practice is discouraged by Google.

Instead the flow is as follows:

* User is asked for permission. The two cases are:
  * Permission is granted.
  * Permission is denied. In this case we ask the OS if we should ask the user again.
   * If the OS says yes, then we ask the user again and provide additional information about why we need that information.
   * If the OS says no, then we disable the feature. We provide no link into OS settings to re-activate permissions.

This is roughly implemented in this PR. In #408 this is polished.

More background on why we should handle it like this: https://github.com/Baseflow/flutter-geolocator/wiki/Breaking-changes-in-7.0.0

If you want to test this build #408

### CODE REVIEW

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt [hunk 1]
```diff
@@ -1,13 +1,9 @@
...
-import app.ehrenamtskarte.backend.auth.database.repos.AdministratorsRepository
 import org.jetbrains.exposed.sql.SchemaUtils
 
 fun setupDatabase() {
     SchemaUtils.create(
         Administrators
     )
-    if (AdministratorEntity.count() == 0L) {
-        AdministratorsRepository.insert("admin@example.com", "admin")
-    }
```
Comment (maxammann) [inline backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt:12]: I silently want to remove this.

Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt:12]: I will take care that we have some login on production and staging.

File: frontend/lib/about/about_page.dart [hunk 2]
```diff
@@ -124,6 +125,23 @@ class AboutPage extends StatelessWidget {
                       launch(
                           "https://github.com/ehrenamtskarte/ehrenamtskarte");
                     }),
+                if (!settings.locationFeatureEnabled)
```
Comment (maxammann) [inline frontend/lib/about/about_page.dart:128]: In #408 this will move to the dev settings.

File: frontend/lib/configuration/settings_model.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import 'package:flutter/foundation.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:ehrenamtskarte/location/determine_position.dart';
+import 'package:geolocator/geolocator.dart';
+
+const hideVerificationInfoPropertyName = "hideVerificationInfo";
+
+class SettingsModel extends ChangeNotifier {
```
Comment (maxammann) [inline frontend/lib/configuration/settings_model.dart:8]: This unifies all persistent settings via ChangeNotifiers.

File: frontend/lib/entry_widget.dart [hunk 4]
```diff
@@ -13,45 +14,49 @@ class EntryWidget extends StatelessWidget {
...
   Widget build(BuildContext context) {
-    return FutureBuilder<bool>(
-        future: isFirstStart(),
+    final settings = Provider.of<SettingsModel>(context);
+    return FutureBuilder<SettingsModel>(
+        future: settings.initialize(),
         builder: (context, snapshot) {
-          if (snapshot.connectionState == ConnectionState.none ||
-              snapshot.connectionState == ConnectionState.waiting) {
+          final settings = snapshot.data;
```
Comment (maxammann) [inline frontend/lib/entry_widget.dart:21]: This is a slight simplification.

File: frontend/lib/location/dialogs.dart [hunk 5]
```diff
@@ -43,3 +43,33 @@ class LocationPermissionDialog extends StatelessWidget {
   }
 }
 
+class RationaleDialog extends StatelessWidget {
```
Comment (maxammann) [inline frontend/lib/location/dialogs.dart:46]: This shows a rationale if permissions are denied once.

File: frontend/lib/map/location_button.dart [hunk 6]
```diff
@@ -23,31 +26,37 @@ class _LocationButtonState extends State<LocationButton> {
...
-                        color: theme.colorScheme.secondary)
-                    : const SmallButtonSpinner(),
-                duration: const Duration(milliseconds: 200),
-              ),
-              onPressed: _status == LocationPermissionStatus.requestFinished
-                  ? _onPressed
-                  : null,
-            )));
+    final settings = Provider.of<SettingsModel>(context);
+    if (settings.locationFeatureEnabled) {
```
Comment (maxammann) [inline frontend/lib/map/location_button.dart:30]: In #408 we show the FAB even though the feature is disabled. This provides a way of reenabling it.

File: frontend/lib/search/location_button.dart [hunk 7]
```diff
@@ -21,49 +23,61 @@ class _LocationButtonState extends State<LocationButton> {
...
-            elevation: 1,
-            onPressed: onPressed,
-            icon: AnimatedSwitcher(
-                child: icon, duration: const Duration(milliseconds: 200)),
-            label: Text(
-              "In meiner Nähe suchen",
-              style: TextStyle(color: theme.hintColor),
-            )));
+    final settings = Provider.of<SettingsModel>(context);
+    if (settings.locationFeatureEnabled) {
```
Comment (maxammann) [inline frontend/lib/search/location_button.dart:39]: We need to remove this button if the feature is permanently disabled.

File: frontend/lib/search/search_page.dart [hunk 8]
```diff
@@ -17,66 +17,72 @@ class SearchPage extends StatefulWidget {
 
 class _SearchPageState extends State<SearchPage> {
   String? _searchFieldText;
-  TextEditingController? _textEditingController;
+  final TextEditingController _textEditingController = TextEditingController();
```
Comment (maxammann) [inline frontend/lib/search/search_page.dart:20]: Only simplifications in this file.

File: frontend/lib/intro_slides/location_request_button.dart [hunk 9]
```diff
@@ -39,17 +51,21 @@ class _LocationRequestButtonState extends State<LocationRequestButton> {
...
       case LocationStatus.always:
         return const ElevatedButton(
           onPressed: null,
-          child: Text("Standort ist bereits freigegeben."),
+          child: Text("Standort ist freigegeben."),
+        );
+      case LocationStatus.deniedForever:
+        return const ElevatedButton(
+          onPressed: null,
+          child: Text("Standortfreigabe ist deaktiviert."),
```
Comment (maxammann) [inline frontend/lib/intro_slides/location_request_button.dart:68]: This case is important and correct. For reference open Google Maps and always deny the permission. If you click the "Show my location on Map"-Button you have no options to directly reenable the feature.

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt (hunk 1 above)
Comment (Schedulaar) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt:12]: We should have some easy and simple way to initialize local setups as well

File: frontend/lib/intro_slides/location_request_button.dart (hunk 9 above)
Comment (Schedulaar) [reply frontend/lib/intro_slides/location_request_button.dart:68]: I disagree. If I deny it forever in Google Maps, at some point, the "enable permission" dialog doesn't show up, but in that case, a snackbar appears a link to the app's settings.
image

File: frontend/lib/location/determine_position.dart [hunk 10]
```diff
@@ -106,32 +110,48 @@ Future<LocationStatus> checkAndRequestLocationPermission(
...
     var permission = await Geolocator.checkPermission();
-    
+
     if (permission == LocationPermission.deniedForever) {
-      // denied forever -> open UI
-      if (await openSettingsToGrantPermissions(context)) {
-        // get new result after the dialog is flo
-        var permission = await Geolocator.checkPermission();
-        return permission.toLocationStatus();
-      }
```
Comment (Schedulaar) [inline frontend/lib/location/determine_position.dart:124]: Imo this is still the right thing to do. At least Google Maps does it the same (more or less)

File: frontend/lib/location/determine_position.dart [hunk 11]
```diff
@@ -93,9 +95,11 @@ Future<RequestedPosition> determinePosition(BuildContext context,
 ///
 /// When the location services are not enabled then it will return
 /// LocationPermission.deniedForever
-Future<LocationStatus> checkAndRequestLocationPermission(
-    BuildContext context,
-    {bool requestIfNotGranted = true}) async {
+Future<LocationStatus> checkAndRequestLocationPermission(BuildContext context,
+    {bool requestIfNotGranted = true,
+    rationale =
+        "Erlauben Sie der App ihren Standort zu benutzen um Akzeptanzstellen ihn Ihrer Umgebung anzuzeigen.",
```
Comment (Schedulaar) [inline frontend/lib/location/determine_position.dart:101]:
```suggestion
        "Erlauben Sie der App Ihren Standort zu benutzen, um Akzeptanzstellen in Ihrer Umgebung anzuzeigen.",
```

File: frontend/lib/location/determine_position.dart [hunk 12]
```diff
@@ -93,9 +95,11 @@ Future<RequestedPosition> determinePosition(BuildContext context,
 ///
 /// When the location services are not enabled then it will return
 /// LocationPermission.deniedForever
```
Comment (Schedulaar) [inline frontend/lib/location/determine_position.dart:97]:
```suggestion
```

File: frontend/lib/entry_widget.dart [hunk 13]
```diff
@@ -13,45 +14,49 @@ class EntryWidget extends StatelessWidget {
...
+                locale: const Locale('de'),
+                home: HomePage(
+                  showVerification: Configuration.of(context).showVerification,
+                ),
+                routes: routes);
+          } else if (snapshot.hasError) {
+            return Container();
+          } else {
             return Container();
           }
```
Comment (Schedulaar) [inline frontend/lib/entry_widget.dart:59]: Weeird

File: frontend/lib/location/dialogs.dart [hunk 14]
```diff
@@ -43,3 +43,33 @@ class LocationPermissionDialog extends StatelessWidget {
...
+        ],
+      ) ,
+      actions: [
+        TextButton(
+            child: const Text("OK"),
+            onPressed: () => Navigator.of(context).pop(true)),
+        TextButton(
+            child: const Text("Nein Danke"),
+            onPressed: () => Navigator.of(context).pop(true))
+      ],
```
Comment (Schedulaar) [inline frontend/lib/location/dialogs.dart:70]:
```suggestion
        children: Text(_rationale)
      ) ,
      actions: [
        TextButton(
            child: const Text("Berechtigung erteilen"),
            onPressed: () => Navigator.of(context).pop(true)),
        TextButton(
            child: const Text("Abbrechen"),
            onPressed: () => Navigator.of(context).pop(true))
      ],
```

**Review by Schedulaar — CHANGES_REQUESTED** [review]
Imo, this PR is too large

File: frontend/lib/entry_widget.dart (hunk 13 above)
Comment (maxammann) [reply frontend/lib/entry_widget.dart]: I handle errors now.

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt (hunk 1 above)
Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/Setup.kt:12]: I created a task: https://github.com/ehrenamtskarte/ehrenamtskarte/issues/410

File: frontend/lib/intro_slides/location_request_button.dart (hunk 9 above)
Comment (maxammann) [reply frontend/lib/intro_slides/location_request_button.dart:68]: This is crazy :D In my google maps version there is no SETTINGS button. I change it like it is now in Google maps in #408

Comment (maxammann) [reply frontend/lib/intro_slides/location_request_button.dart:68]: But I'll leave the intro slides as is. No need to allow the user again to give permissions, after he denied them for 2 times.

**Comment by elKei24** [comment]
In the APK you posted in Mattermost, after already permitting location access in the intro slides, I still have to enable the location-related features in the map/search additionally. At the next app start, this is not the case any more. Is this intended?

https://user-images.githubusercontent.com/13063245/141822340-ab184528-06c5-4434-8e2e-76b3cc698c5c.mp4

**Comment by maxammann** [comment]
> In the APK you posted in Mattermost, after already permitting location access in the intro slides, I still have to enable the location-related features in the map/search additionally. At the next app start, this is not the case any more. Is this intended?
> 
> https://user-images.githubusercontent.com/13063245/141822340-ab184528-06c5-4434-8e2e-76b3cc698c5c.mp4

This behavior is not new. I think this is not a huge issue. We maybe should open a bug though. Thanks for testing! :)

**Pull request merged**